pkg/pillar/cmd/zedagent/localinfo.go

```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 			continue
 		}
 		// Try to find the application instance.
-		ais := findAppInstance(ctx, appUUID, displayName)
-		if ais == nil {
+		appInst := findAppInstance(ctx, appUUID, displayName)
+		if appInst == nil {
 			log.Warnf("Failed to find app instance with UUID=%s, displayName=%s",
 				appUUID, displayName)
 			continue
 		}
-		appUUID = ais.UUIDandVersion.UUID
+		appUUID = appInst.UUIDandVersion.UUID
+		if _, duplicate := publishApps[appUUID.String()]; duplicate {
+			log.Warnf("Multiple commands requested for app instance with UUID=%s",
+				appUUID)
+			continue
+		}
+
 		command := types.AppCommand(appCmdReq.Command)
-		appCmd := ctx.localAppCommands.LookupByAppUUID(appUUID)
-		if appCmd != nil {
-			// Entry for this app already exists.
-			if appCmd.Command == command &&
-				appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
-				// already accepted
-				continue
+		appCmd, hasLocalCmd := ctx.localCommands.AppCommands[appUUID.String()]
+		if !hasLocalCmd {
+			appCmd = &types.LocalAppCommand{}
+			if ctx.localCommands.AppCommands == nil {
+				ctx.localCommands.AppCommands = make(map[string]*types.LocalAppCommand)
 			}
-			appCmd.Command = command
-			appCmd.LocalServerTimestamp = appCmdReq.Timestamp
-			appCmd.DeviceTimestamp = time.Now()
-			appCmd.Completed = false
-			cmdChanges = true
+			ctx.localCommands.AppCommands[appUUID.String()] = appCmd
+		}
+		if appCmd.Command == command &&
+			appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
+			// already accepted
 			continue
 		}
-		// Add new entry.
-		ctx.localAppCommands.Cmds = append(ctx.localAppCommands.Cmds,
-			types.LocalAppCommand{
-				AppUUID:              appUUID,
-				Command:              command,
-				LocalServerTimestamp: appCmdReq.Timestamp,
-				DeviceTimestamp:      time.Now(),
-				Completed:            false,
-			})
+		appCmd.Command = command
+		appCmd.LocalServerTimestamp = appCmdReq.Timestamp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		appCounters, hasCounters := ctx.localCommands.AppCounters[appUUID.String()]
```

Comment:
It feels like we should be able to have an abstration/function here which takes care of updating the appInst+volumeref+volume and republish. Presumably it wouldn't need anything but appCmd.Command and it wouldn't modify appCmd?

Reply:
Done - added a new function `triggerLocalCommand`